packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts

```diff
@@ -324,26 +346,61 @@ function isValidEventForCustomElement(event: Event): boolean {
     const target = eventTargetGetter.call(event);
     const currentTarget = eventCurrentTargetGetter.call(event);
     const { composed } = event;
-    return (
-        // it is composed, and we should always get it, or
-        composed === true ||
-        // it is dispatched onto the custom element directly, or
-        target === currentTarget ||
+
+    let shouldInvoke = false;
+
+    if (isTrue(composed)) {
+        shouldInvoke = true;
+    } else if (target === currentTarget) {
+        // Address the possibility that `target` is a shadow root
+        shouldInvoke = !eventsDispatchedDirectlyOnShadowRoot.has(event);
+    } else {
         // it is coming from a slotted element
-        isChildNode(
+        shouldInvoke = isChildNode(
             getRootNodeHost(
                 target as Node /* because wrap on shadowRoot */,
                 GET_ROOT_NODE_CONFIG_FALSE
             ) as Element,
             currentTarget as Node
-        )
-    );
+        );
+    }
+
+    return shouldInvoke;
+}
+
+function isValidEventForShadowRoot(event: Event): boolean {
```

Comment:
What about renaming this function to `shouldInvokeShadowRootListener`? And use the same name for the function above?

Reply:
https://github.com/salesforce/lwc/pull/2117/commits/1a4efebd4914c6dcbf0c75b4b2da577439478aea